### ISSUE TITLE
Request: add beep command

### ISSUE THREAD

**Issue by steve6375** [chenall/grub4dos#175]
beep [Frequency Hz] [Duration mS] [Frequency Hz] [Duration mS]... etc.

e.g.  Usage:
beep 1000 500                            (PC speaker 1kHz for 1/2 second)
beep 1000 500 0 1000 500 500  (PC Speaker 1kHz for 1/2 second, 1 sec silence, 500Hz for 1/2 second)
beep < mytune.txt                      (space\tab\cr\lf are ignored)

similar to this for nircmd
nircmd beep 311 250
nircmd beep 415 450
nircmd beep 311 150
nircmd beep 415 250
nircmd beep 311 150
nircmd beep 415 150
nircmd beep 466 250
nircmd beep 523 600
nircmd beep 415 300

https://wiki.osdev.org/PC_Speaker

or maybe implement as  grub4dos executable?

**Comment by yaya2007** [comment]
The opening and closing time can only be set.
Cannot set frequency, It's fixed. 44100Hz.

Also inform:
2018-04-23 version, You need to increase the menu item !BAT to use batch functions, such as goto: Label .

**Comment by steve6375** [comment]
This DOS program can play a wav or PCM file just using the PC Speaker.
https://github.com/miwelc/Internal-PC-Speaker-player--MS-DOS-Dos-Box-

**Comment by yaya2007** [comment]
Try,please.
title beep
!BAT
:a
beep 2000 550 && beep 1000 550 && beep 500 550 && beep 0 55 && beep 500 550 && beep 1000 550 && beep 2000 550 && beep 0 55 && goto :a

Terminate by any key.

**Comment by steve6375** [comment]
Great Work!  Beep seems to work fine and also frequency 0 supported too!

However, the example menu you gave does not loop and says 'cannot find the batch label specified - a'
???
Also debug info printed when key pressed.

**Comment by steve6375** [comment]
Demo tunes  :-)
[medley.zip](https://github.com/chenall/grub4dos/files/1988822/medley.zip)

**Comment by yaya2007** [comment]
Is it accurate to describe ?
  "beep [--loop] FREQUENCY DURATION FREQUENCY DURATION ...",
  "The unit of frequency is Hz, the unit of duration is ms.\n"
  "The parameters are not more than 126."

Duration: 50 ms=1 ticking

**Comment by steve6375** [comment]
Max. 126 beep parameters per command line.

How about loop=2 or loop=0 for infinite loop?

**Comment by yaya2007** [comment]
[grldr.rar.txt](https://github.com/chenall/grub4dos/files/1991329/grldr.rar.txt)

**Comment by yaya2007** [comment]
"beep [--loop=N] FREQUENCY DURATION FREQUENCY DURATION ...",
  "The unit of frequency is Hz, the unit of duration is ms.\n"
  "Max. 63 beep parameters per command line."

**Comment by steve6375** [comment]
If a key is pressed to stop the loop, the key scan code is not emptied from the keyboard buffer, so it is passed onto grub4dos.
e.g.
in command console
beep --loop=99 1000 1000
if ESC is pressed then command console exits back to menu.
If L is pressed, L is sent to command prompt.

I think the key press should be absorbed?

**Comment by steve6375** [comment]
P.S.  --loop=0 only plays once. How to do infinite loop?

**Comment by yaya2007** [comment]
"beep [--loop=N] FREQUENCY DURATION FREQUENCY DURATION ...",
  "The unit of frequency is Hz, the unit of duration is tick(55ms).\n"
  "Max. 126 beep parameters per command line."

title beep
echo MoLiHua
beep 330 8 330 4 392 4 440 4 522 4 522 4 440 4 392 8 393 4 440 4 392 16
beep 330 8 330 4 392 4 440 4 522 4 522 4 440 4 392 8 392 4 440 4 392 16
beep 392 8 392 8 392 8 330 4 392 4 440 8 440 8 392 16
beep 330 8 293 4 330 4 392 8 330 4 293 4 261 8 261 4 293 4 261 16
beep 330 4 293 4 261 4 330 4 293 12 330 4 392 8 440 4 522 4 392 16
beep 293 8 330 4 392 4 293 4 330 4 261 4 220 4 196 16 220 8 261 8
beep 293 12 330 4 261 4 293 4 261 4 220 4 196 24

**Comment by steve6375** [comment]
No!
This makes it very hard to write tunes if duration is in ticks. nircmd and powershell use ms.
All tunes are in ms not ticks.
Why did you change it?
Please change back to ms for DURATION?

Can you support --loop=0 for infinite loop please? Add parameter for support for ticks e.g. --ticks

Also in help text add 
N=0 (inifinite),1-65535 (max)
All numbers are Integers only.

Thanks

**Comment by yaya2007** [comment]
In milliseconds, it is converted into a tick inside. For example, 50_99 milliseconds, all converts to 1 ticks.

**Comment by steve6375** [comment]
In previous version
beep 0 1000    delayed for 1 second
now in latest version it delays for 55 seconds!

Why did you change it?

Duration should be in units of ms.

**Comment by yaya2007** [comment]
There is no timer inside, it is the ticking of BIOS. All the milliseconds turn to the tick first. That is: MS / 55

**Comment by yaya2007** [comment]
Or the precision of the duration is not 1ms, but 55ms.

**Comment by steve6375** [comment]
I know and understand.
But the duration parameter should be in units of ms, as in first version.
You have changed it to ticks, but parameter should be ms.
E.g. beep 0 1000 should delay for 1 second.

**Comment by steve6375** [comment]
If parameter x in ms then
Ticks = (x + 27) / 55

**Comment by yaya2007** [comment]
A timer is made, and the exact 1ms can be made. 

  "beep [--start|--mid|--end] [--loop=N] FREQUENCY DURATION FREQUENCY DURATION ...",
  "FREQUENCY: Hz. DURATION: ms. Max: 126 beeps per command.\n"
  "N: 0-254. 255 is an infinite loop(Press F2 key terminate).\n"
  "When the syllable is a lot, can be written in different lines.\n"
  "The use of [--start|--mid|--end] specifies."

Examples (in the menu) :
timeout 15

beep --start 330 440 330 220 392 220 440 220 522 220 522 220 440 220 392 440 393 220 440 220 392 880
beep --mid 330 440 330 220 392 220 440 220 522 220 522 220 440 220 392 440 392 220 440 220 392 880
beep --mid 392 440 392 440 392 440 330 220 392 220 440 440 440 440 392 880
beep --mid 330 440 293 220 330 220 392 440 330 220 293 220 261 440 261 220 293 220 261 880
beep --mid 330 220 293 220 261 220 330 220 293 660 330 220 392 440 440 220 522 220 392 880
beep --mid 293 440 330 220 392 220 293 220 330 220 261 220 220 220 196 880 220 440 261 440
beep --end --loop=255 293 660 330 220 261 220 293 220 261 220 220 220 196 1320

The infinite cycle beep sound and the infinite cycle animation can match well. 

The delay parameter of animation can be changed from tick to millisecond. Is it a direct change or set a switch?

**Comment by steve6375** [comment]
Wow - very impressive work!

Some issues:
1. How do you stop tune if --loop=255 and --start --mid --end is used? Another beep command has no effect. e.g. beep 0 0  could cancel playing beep?
2. Even if boot to MSDOS from grub4dos, it still plays a continuous beep!
3. If load 2nd grub4dos menu,  beep play sequence is stretched during execution of commands.

Suggest any key press stops  beep if --loop and --start --mid --end is used but key scan code is not absorbed but passed on to grub4dos?  e.g.  menu with animation and beep playing, user presses ENTER or hotkey and beep stops and menu entry executed.

i.e.
beep --loop=10 300 300 400 400          - user press key to stop, key is absorbed

beep --start 330 440
beep --mid 392 220
beep --end --loop=255 330                  - user presses any key to stop, key is not absorbed


re. Animation - I think ticks is OK, no need for change. I have not needed more resolution or faster animation. It must be compatible with previous versions.

**Comment by yaya2007** [comment]
255 is an infinite loop(Press F2 key terminate)

**Comment by steve6375** [comment]
F2 does not work if animation playing - animation cannot be started again.  I think any key should stop play if --end used.

Also
--loop=0 plays once, --loop=1 plays twice
I suggest change so --loop=0 stops play, --loop=1 plays once?

Also would be nice to have synchronous play always, not just if loop=255 - i.e. use of --start --mid --end always plays synchronously, so can have menu with continuous animation and tune played 3 times in loop

e.g.
beep --start 300 300 0 200
beep --end --loop=3  200 300 0 200
will play synchronously whilst menu is showing and animation but any key press stops beep and clears beep command buffer.



re. help beep
Max: 126 **notes** per command.

N: 0-255. 0 stops synchronous play, 255 is continuous play (any key stops play).
synchronous play:  one --start, 0 or more --mid and one --end lines can be used.

**Comment by steve6375** [comment]
P.S. if multiple --end is used, it adds extra beeps, e.g.

beep --start 200 200 0 300
beep --end 300 300 0 300
... plays two beeps
beep --end 400 400 0 400
...plays 3 beeps

--end should clear previous buffer?

**Comment by yaya2007** [comment]
any key stops play:
You can do it without an infinite loop.

Press the F2 key to terminate the animation,The second press F2 key starts playing, it's a byproduct.
You need this byproduct?

"beep --start 200 200 0 300
 beep --end 300 300 0 300
 ... plays two beeps
 beep --end 400 400 0 400
 ...plays 3 beeps"
beep 200 200 0 300
beep 300 300 0 300
beep 400 400 0 400

**Comment by steve6375** [comment]
If I have continuous animation in menu and --loop=255 then:
First press F2 removes beep and animation
2nd press F2 does nothing

I want any key press to stop synchronous beep.
Synchronous beep  = --start --mid --end usage

e.g. I want this functionality

beep --loop=3 300 3000 400 3000            - asynchronous, any key press stops and kbd buffer is emptied

beep 300 300 440 400 || goto :EOF
beep 441 300 445 300 || goto :EOF
etc.
plays long tune, key press aborts, kbd buffer emptied

beep --start 300 3000 400 3000
beep --end 440 400                       - synchronous play once whilst menu and animation active
any key press stops beep and clears beep buffer, kbd buffer is NOT emptied

beep --start 300 3000 400 3000
beep --end --loop=3 440 400 0 3000     - synchronous play whilst menu and animation active
any key press stops beep and clears beep buffer, kbd buffer is NOT emptied

--loop=255 - same as --loop=3 but continuous synchronous play

**Comment by steve6375** [comment]
P.S. In case of problems when booting to DOS and beep is playing, TSR interrupt for beep should be removed when 'boot' is executed by grub4dos.

**Comment by yaya2007** [comment]
This has been dealt with. Stop animation and sound when the menu item is executed.

**Comment by yaya2007** [comment]
"animation active"
How to do it。

**Comment by yaya2007** [comment]
The use of (-start --mid --end） is equivalent to one line. In one line, an arbitrary key is withdrawn.

**Comment by yaya2007** [comment]
What is KDB

**Comment by steve6375** [comment]
There should be a loop counter in TSR interrupt routine.
If user specifies --start/mid/end to invoke beep TSR interrupt routine.
If no --loop specified then TSR_loop=1, if --loop=255 then TSR_loop=infinite, else TSR_loop=1-254.
If no --start/mid/end then just play beep (no TSR). e.g. 

beep --loop=255 300 300  (no TSR interrupt routine, just play indefinitely)

beep --start 0 1
beep --end 300 300        (TSR interrupt used)

beep --start 0 1
beep --end --loop=2  300 300 (TSR interrupt used)

beep --start 0 1
beep --end --loop=255  300 300 (TSR interrupt used)

**Comment by steve6375** [comment]
KBD=keyboard

**Comment by yaya2007** [comment]
(-start --mid --end) just merge multiple rows into one row. Don't overuse it. No more discussion in the future. Let's talk about one line of beep.

**Comment by steve6375** [comment]
OK, so if use --loop, can it always use an interrupt routine please?

I want to play a tune synchronously as menu loads and displays animation graphics.

Tune can be made to play once (--loop=1) or 3 times (--loop=2-254) or infinite (--loop=255) or stopped if already playing (--loop=0) or by any other beep command.

**Comment by yaya2007** [comment]
"beep [--start|--mid|--end] [--play=N] [--unshared] FREQUENCY DURATION FREQUENCY DURATION ...",
  "FREQUENCY: Hz. DURATION: ms. Max: 126 notes per command.\n"
  "N: 0-255. 0 is stop play, 255 is continuous play (F2 key stops play).\n"
  "When the syllable is a lot, can be written in different lines.\n"
  "The use of [--start|--mid|--end] specifies."

title test
splashimage --animated=0x83=5=3=230=150 (3)/E2BRotate_frame_0001.bmp
beep --play=3 440 500 440 500 440 500 349 350 523 150 440 500 349 350 523 150

**Comment by steve6375** [comment]
Animation issue: 
If press F2 then animation stops
If press F2 again, animation does not start.
If load 2nd grub4dos menu using configfile, animation is not present in new menu
if run any menu entry, animation disappears

beep issues:
beep 200 200                                              beeps twice
beep --unshared --play=100 1000 0          cannot stop beep  - is 0 duration infinite?
beep --unshared --play=1 1000 200 0 200  plays 2 beeps
beep 311 250 415 450 311 150 415 250  311 180  415 180  466 250  523 600  415 300  plays extra note at beginning and plays tune twice

!BAT
:BEGIN
echo Axel Foley (press a key to abort)
beep 659 460 784 340 659 230 659 110 || goto :EOF
beep 880 230 659 230 587 230  || goto :EOF
beep 659 460 988 340 659 230 659 110 || goto :EOF
beep 1047 230 988 230 784 230 659 230 || goto :EOF
beep 988 230 1318 230 659 110 587 230 || goto :EOF
beep 587 110 494 230 740 230 659 460 || goto :EOF
beep 0 500 && goto :BEGIN

This no longer works

I prefer any key press to stop beep because user will not know to press F2 and sound can be very loud and is annoying. Can we have --anykey parameter so sound stops with any key press please? I do not want F2 to control beep at all, can it be removed please? If user wants to listen to music they simply do not press a key.

**Comment by yaya2007** [comment]
!BAT
 echo Axel Foley (press a key to abort)
 beep  --unshared 659 460 784 340 659 230 659 110 || goto :EOF
 beep --unshared  880 230 659 230 587 230 || goto :EOF
 beep --unshared  659 460 988 340 659 230 659 110 || goto :EOF
 beep  --unshared 1047 230 988 230 784 230 659 230 || goto :EOF
 beep  --unshared 988 230 1318 230 659 110 587 230 || goto :EOF
 beep  --unshared 587 110 494 230 740 230 659 460 || goto :EOF
 beep  --unshared 0 500 && goto :BEGIN

**Comment by steve6375** [comment]
rather than --unshared,  how about  beep --wait ?

**Comment by steve6375** [comment]
```
beep --unshared 392 350 0 100 392 350 0 100 392 350 0 100 311 250 0 100 466 25 0 100 392 350 0 100 311 250 0 100 466 25 0 100 392 700 0 100 587 350 0 100
beep --unshared 587 350 0 100 
```
This does not play correctly (first line plays twice)

**Comment by yaya2007** [comment]
any key stops play

**Comment by steve6375** [comment]
Almost working now!

animation is removed still even if press cursor key in menu,

beep --play=10 400 20 500 20 600 20 700 20 800 20 900 20 1000 20 1100 20 1200 20 1300 20 1400 20 1500 20 1600 20

plays extra long note at start

**Comment by steve6375** [comment]
OK - re. extra long note.
This is caused by processor.
If I use a menu entry to run a .g4b batch file then first note is stretched by CPU activity.

Can I suggest you add a 1 second delay before beep command is executed to avoid this affect?

If user adds delay then we get a delay inside the loop which we do not want.
beep --play=10 **0 1000** 400 20 500 20 600 20 700 20 800 20 900 20 1000 20 1100 20 1200 20 1300 20 1400 20 1500 20 1600 20

So you need to add delay at start only (unless --wait is used).

run this batch file from a menu entry for demo
!BAT
echo Alarm!
beep --play=10  400 20 500 20 600 20 700 20 800 20 900 20 1000 20 1100 20 1200 20 1300 20 1400 20 1500 20 1600 20

**Comment by yaya2007** [comment]
It's too loud and a bit scary.

**Comment by yaya2007** [comment]
I look up the error.

**Comment by steve6375** [comment]
[Tunes.zip](https://github.com/chenall/grub4dos/files/2016837/Tunes.zip)

Here are my test files. Try Alarm.g4b called from menu to see issue.

**Comment by steve6375** [comment]
another issue. Pressing a letter key does not stop the tune if there is a menu entry present that starts with the same letter. The routine that gets key press and looks for menu entry that starts with that letter to select it in the menu is soaking up the key press.

e.g. if there is a menu entry for "Ubuntu" then pressing U does not stop tune, but pressing Z does stop tune if no menu entry beginning with Z.

**Comment by yaya2007** [comment]
"Here are my test files. Try Alarm.g4b called from menu to see issue."

**Comment by yaya2007** [comment]
"e.g. if there is a menu entry for "Ubuntu" then pressing U does not stop tune, but pressing Z does stop tune if no menu entry beginning with Z."
The problem is not reproduced.
Do you use hot keys?

**Comment by steve6375** [comment]
Hotkey utility is loaded. Try easy2boot.

**Comment by yaya2007** [comment]
I loaded Hotkey, and I didn't make a mistake.

**Comment by yaya2007** [comment]
[grldr.rar.txt](https://github.com/chenall/grub4dos/files/2017362/grldr.rar.txt)

**Comment by steve6375** [comment]
delay added seems to work :-)

Please make new USB flash drive using attached file
[beeptestE2B.zip](https://github.com/chenall/grub4dos/files/2017402/beeptestE2B.zip)
![image](https://user-images.githubusercontent.com/1469350/40242039-6d8e874c-5ab4-11e8-83bc-31c8152cba08.png)

Observe rotating animated icon
Now press D key - animation will continue.
Now press U or H or B - animation will continue
Now press Z or cursor down - animation will stop

Now play tune using menu - 2 AxelFoley
When tune starts  press D or U or H or B - tune continues to play

**Comment by steve6375** [comment]
Also GuitarTuner demo now not working correctly or StarWarsLong or SuperMario or TestrisWait????

**Comment by steve6375** [comment]
Help text:
was:
"When the syllable is a lot, can be written in different lines.\n
"The use of [--start|--mid|--end] specifies."
suggest:
"For long tune, use --start for first phrase, zero or more --mid for middle phrases\n"
"and --end for last phrase (with optional --play=N and\or --unshared).\n"

P.S. I still think --wait is a much better keyword to use than --unshared.

**Comment by yaya2007** [comment]
about "--unshared":
unshared, Exclusive Time, Do not display the menu, Animation and music can't coexist.
How do you use a short word?

**Comment by steve6375** [comment]
Thanks! That is playing better now.

Just the bug of a key press stops animation to fix. This is nothing to do with beep command.
Animation should not stop unless F2 is pressed. If F2 is pressed again, animation should start.
Any other key should not affect animation.

re. --wait - users may not know what --unshared means, but if you use --wait as the parameter instead of --unshared it is easy to understand because grub4dos 'waits' whilst the tune is played.

"beep [--start|--mid|--end] [--play=N] [--wait] FREQUENCY DURATION FREQUENCY DURATION ..."

**Comment by yaya2007** [comment]
I imagine playing both music and animation at the same time. When one stops, the other will stop at the same time. It is also synchronous.

**Comment by steve6375** [comment]
No. Animation should be kept on all menus (menu1, menu2, menu3) etc. same as previous versions.
Animation  stopped and started by F2 only.

beep stopped by any key (because it will become very annoying!)

**Comment by steve6375** [comment]
[SuperMarioss.zip](https://github.com/chenall/grub4dos/files/2019418/SuperMarioss.zip)
This corrupts display (!!) and does not play properly. Should be long beep at end - it also repeats when it should not.
Is there a limit to  beep buffer which has been exceeded?

**Comment by yaya2007** [comment]
313 > 126

**Comment by steve6375** [comment]
Ah - I thought that was a command line limitation!

**Comment by yaya2007** [comment]
Have a try 
"beep [--start|--mid|--end] [--play=N] [--wait] FREQUENCY DURATION FREQUENCY DURATION ...",
[grldr.rar.txt](https://github.com/chenall/grub4dos/files/2019548/grldr.rar.txt)

**Comment by steve6375** [comment]
looking really good! Great job!
I will fully test some more later. Thank you for all your hard work :-)

**Comment by steve6375** [comment]
P.S. Instead of   Max: 126 notes per command,
change to
Max: 126 notes.

**Comment by steve6375** [comment]
So far testing is looking very good :-) . No problems found.

However, powershell and nircmd beep commands can be used like this
beep 300 200
beep 400 200
beep 500 200

whereas if this is done in grub4dos it does not work (because you have to use --wait with each line).
I think to be more logical and easier for the user to understand, the grub4dos beep command should use --wait by default and the user should specify --nowait if they want synchronous play, then consecutive beep commands will work as expected by the user. i.e.

"beep [--start|--mid|--end] [--play=N] [--nowait] FREQUENCY DURATION FREQUENCY DURATION ...",

**Comment by yaya2007** [comment]
"beep [--start|--mid|--end] [--play=N] [--nowait] FREQUENCY DURATION FREQUENCY DURATION ..."
[grldr.rar.txt](https://github.com/chenall/grub4dos/files/2020004/grldr.rar.txt)

**Comment by steve6375** [comment]
Looks very good now. Thanks a lot - great work!

The only problem I have found is if large jpg animation is playing and with delay of 1 (fastest) then it plays a tune slowly instead of at correct speed. Presumably this is due to the speed of writing the animation frames to the screen. However this is not a great problem.
Perhaps if the timer resolution was increased to 5ms instead of 1ms it would work better and give 5ms for the animation frame to be written to the screen? I doubt if 1ms resolution is really needed anyway?

**Comment by steve6375** [comment]
[E2B_animation_slow_demo.zip](https://github.com/chenall/grub4dos/files/2020040/E2B_animation_slow_demo.zip)

**Comment by yaya2007** [comment]
The parameters of the animation delay=0x3? 
Play animation alone , or play animation and music at the same time , No difference was found .

**Comment by yaya2007** [comment]
Jpg image decoding time is long (not as fast as BMP image). This time should also be used as part of the delay time. If 1 ticks can not be expressed, you can use milliseconds. Look at the help. For example, 25ms can use = 25:ms =

**Comment by steve6375** [comment]
delay for splashimage animation is 1  - see \_ISO\MyE2B,cfg
set ANIMATE=0x10=1=%last%=575=225 (fd3)/E2B/sphereR_frame_0001.JPG
e.g. AxelFoley plays much faster when F2 is pressed to turn off animation (on my HP G62 laptop).

I tried
set ANIMATE=0x10=55:ms=%last%=575=225 (fd3)/E2B/sphereR_frame_0001.JPG
and it is the same - tune is playing slower and takes longer to finish.
does it take 55ms or more for jpg decode?

**Comment by steve6375** [comment]
1 tick delay for animation
with animation - 3 loops of Axel Foley takes 20.6 seconds
no animation -  17.9

175:ms delay
with animation - 3 loops of AxelFoley takes 18.5 seconds
no animation - 17.9

**Comment by yaya2007** [comment]
set ANIMATE=0x10=55:ms=%last%=575=225 (fd3)/E2B/sphereR_frame_0001.JPG
Cycle once 5'17 seconds

set ANIMATE=0x10=1:ms=%last%=575=225 (fd3)/E2B/sphereR_frame_0001.JPG
Cycle once 7'37 seconds

**Comment by steve6375** [comment]
Even with frame delay of 175 ms there is still evidence of interrupts being missed. Presumably interrupts are disabled in frame update routine the whole time and so some 1ms interrupts are being missed. Can interrupts be re-enabled during frame update?

**Comment by yaya2007** [comment]
The above is: Test results in a virtual machine.

Real machine test:
1ms is too fast 
10ms is a bit fast 
20ms is better 

In addition, the transparent background has a color block.
The background color of the animation is best used in black.

**Comment by steve6375** [comment]
sure, but even when using delay of 175ms for animation, the tune plays slower.
So it is not just a question of the frames taking too long to display.
It looks like interrupts are missed even if there is plenty of time for 1 frame to be written to display.

**Comment by yaya2007** [comment]
Infinite loop animation is now loaded into memory, and there is no decompression delay time. 
[grldr.rar.txt](https://github.com/chenall/grub4dos/files/2042380/grldr.rar.txt)

**Comment by steve6375** [comment]
E2B loads the frames into memory anyway.
It loads a floppy disk .ima file into memory and then uses (fd3)/frame001.jpg  etc. as the frame path.

There is still a delay
```
set ANIMFD3=/_ISO/sphereR.ima
set last=90
set ANIMATE=0x90=2=%last%=615=160 (fd3)/E2B/sphereR_frame_0001.BMP
if not exist DONEMENU call /_ISO/docs/Tunes/AxelFoley_Loop.g4b

```
AxelFoley 3 loops = 22 seconds with animation playing, 17 seconds with no animation.
It looks to be only a bit better.
I think the previous version could be fixed if you enabled interrupts whilst writing the animation frame to display memory. The problem is that interrupts are being missed because they have been turned off whilst the frame is being written.

Also there is graphics memory corruption! See picture.
![image](https://user-images.githubusercontent.com/1469350/40584733-01810c6e-619e-11e8-8d0d-19192bd355cd.png)

**Comment by yaya2007** [comment]
Infinite loop: animated images are printed to memory in advance, and only memory is copied during playback. You should test the infinite loop. There is no change in the number of times. When you play, you need to decompress the JPG image, which will take more time.

**Comment by yaya2007** [comment]
615+198＝813
813＞800

**Comment by yaya2007** [comment]
Infinite cycle animation, delay 20ms, to see if there is any delay in music.

**Comment by steve6375** [comment]
813＞800 - my mistake - sorry!

Results for 2018-05-27 version
sphere_bmp.ima - delay 2 - non-transparent - axel foley infinite, 3 loops = 22 seconds
sphere_bmp.ima - delay 2 - transparent - axel foley infinite, 3 loops = 22 seconds
sphere_jpg.ima - delay 2 - axel foley infinite, 3 loops = 22 seconds
(no animation) + axel foley infinite, 3 loops = 18 seconds

So similar times to previous version.
I still think interrupts are being lost due to interrupt routine not enabling interrupts?

**Comment by steve6375** [comment]
Results for 2018-05-20 were better!
sphere_bmp.ima - delay 2 - transparent - axel foley infinite, 3 loops = 20 seconds
sphere_jpg.ima - delay 2 - axel foley infinite, 3 loops = 19.5 seconds
(no animation) + axel foley infinite, 3 loops = 18 seconds

Maybe if you used 5ms interrupt instead of 1ms then much fewer interrupts would be lost?
Just adjust user parameters by adding 2 and dividing by 5.

**Comment by yaya2007** [comment]
Results for 2018-05-28 version
sphere_jpg.ima - transparent - delay 2 - axel foley infinite, 3 loops = 20.6 seconds
sphere_jpg.ima - transparent - delay 20ms - axel foley infinite, 3 loops = 20.6 seconds

Results for 2018-05-20 version
sphere_jpg.ima - transparent - delay 20ms - axel foley infinite, 3 loops = 23.6 seconds
[grldr.rar.txt](https://github.com/chenall/grub4dos/files/2044016/grldr.rar.txt)

**Comment by yaya2007** [comment]
Results for 2018-05-28 version
sphere_jpg.ima - transparent - delay 2 - axel foley infinite, 3 loops = 20.6 seconds
sphere_jpg.ima - transparent - delay 20ms - axel foley infinite, 3 loops = 20.6 seconds

Results for 2018-05-20 version
sphere_jpg.ima - transparent - delay 20ms - axel foley infinite, 3 loops = 23.6 seconds

**Comment by yaya2007** [comment]
Results for 2018-05-28 version
sphere_jpg.ima - transparent - delay 2 - axel foley infinite, 3 loops = 20.6 seconds
sphere_jpg.ima - transparent - delay 20ms - axel foley infinite, 3 loops = 20.6 seconds

Results for 2018-05-20 version
sphere_jpg.ima - transparent - delay 20ms - axel foley infinite, 3 loops = 23.6 seconds

**Comment by steve6375** [comment]
HP laptop G62, grub4dos 0.4.6a 2018-05-28, axel foley 3 loops, sphere gif in bmp of jpg format 90 frames, delay 2 ticks (=110ms)

bmp transparent       22 seconds
bmp non-transparent 22 seconds
jpg transparent         22 seconds
jpg non-transparent 22 seconds
no animation = Axel foley 3 loops plays in **18 seconds**

Bug: if go into grub4dos console (shift+c) animation is still playing in console!

**Comment by steve6375** [comment]
[Sphere_IMA_images.zip](https://github.com/chenall/grub4dos/files/2044175/Sphere_IMA_images.zip)

```
#BMP 198 wide
set ANIMFD3=/_ISO/spherebmp.ima
set last=90
# 1st param - use 0x90 for looping+transparent background, use 0x10 for opaque background
# 2nd param is delay (1\18th second), 3rd is last frame number, 4th is x pos, 5th is y pos (in pixels), 6th is path to first frame
set ANIMATE=0x10=2=%last%=590=160 (fd3)/E2B/sphere_frame_0001.BMP

set ANIMFD3=/_ISO/spherejpg.ima
set last=90
# 1st param - use 0x90 for looping+transparent background, use 0x10 for opaque background
# 2nd param is delay (1\18th second), 3rd is last frame number, 4th is x pos, 5th is y pos (in pixels), 6th is path to first frame
set ANIMATE=0x90=2=%last%=590=160 (fd3)/E2B/sphere_frame_0001.JPG

```

**Comment by steve6375** [comment]
bug #2: if play non-transparent jpg or bmp then last frame (or maybe first frame) shows as transparent frame (can see background) for one frame.

**Comment by steve6375** [comment]
Sorry! I Didn't mean to close this!
Are you still working on delay issue?
If not, I am happy for this to become a new release version once bugs are fixed as mentioned above or revert to 2018-05-20 version?
If you are still working on delay issue, I am happy to test further.

**Comment by yaya2007** [comment]
It's too loud. I thought you gave up beep. 

The transparent background algorithm is modified. 
There is no gray interference on the JPG image. 

Take the version of 2018-05-20 as a template.

**Comment by steve6375** [comment]
it's crashes/hangs - no menu??? bad compile???

**Comment by yaya2007** [comment]
[grldr.rar.txt](https://github.com/chenall/grub4dos/files/2080730/grldr.rar.txt)

**Comment by steve6375** [comment]
Thanks1
but bug - If stop animation using F2 and then play tune, the animation re-appears!

**Comment by yaya2007** [comment]
[grldr.rar.txt](https://github.com/chenall/grub4dos/files/2084501/grldr.rar.txt)

**Comment by steve6375** [comment]
Thanks. Thats better. Will put in E2B v1.99h Beta and see if anyone has any issues...
https://goo.gl/vdPuZX

**Comment by yaya2007** [comment]
Unable to open a web site

**Comment by steve6375** [comment]
Thanks for all your hard work on this feature :-)
